### PR TITLE
[EDNA-14] Add server: Castor

### PR DESCRIPTION
**Pull request by mkaito**

### CODE REVIEW

**Review by balsoft — APPROVED** [review]

File: servers/castor/edna.nix
Comment (gromakovsky) [inline servers/castor/edna.nix:57]:
```suggestion
        image = "ghcr.io/serokell/edna-frontend";
```

Maybe I misunderstand something, but I guess it should be `frontend`.

File: servers/castor/edna.nix
Comment (gromakovsky) [inline servers/castor/edna.nix:73]: Who is the consumer of these 3 env variables?

(probably a stupid question, I just didn't deal with frontend before)

File: servers/castor/edna.nix
Comment (gromakovsky) [inline servers/castor/edna.nix:87]: In https://github.com/serokell/edna/pull/8/files it's done a bit differently
image

I. e. there we expect a connection string where you can specify any parameters you want.

I think for development we will just use some pre-defined config, so we can implement any of these options (one conn string or separate options for `host`, `database`, etc.). What do you prefer?

File: servers/castor/edna.nix
Comment (gromakovsky) [inline servers/castor/edna.nix:78]: The format of the config file is quite unstable and the whole concept of this yaml config is debatable, we may even replace it by something else (there were talks about that). For this reason it's quite sad that the config is written in this repo, it means that whenever we update the config structure we have to make a PR to this repo. Is it possible to move more stuff into `edna` repo (here: https://github.com/serokell/edna/pull/11)?

If it doesn't make sense to move anything there, I guess we can live with it.

File: servers/castor/edna.nix
Comment (mkaito) [reply servers/castor/edna.nix:18]: 80 is only used to serve the ACME challenge. All normal traffic is redirected to 443.

File: servers/castor/edna.nix
Comment (mkaito) [reply servers/castor/edna.nix:57]: No, you're right. This silly mistake had me running in circles for half a day. I had a fix locally, but forgot to push. Nice catch.

File: servers/castor/edna.nix
Comment (mkaito) [reply servers/castor/edna.nix:73]: They are used in the Caddyfile: https://github.com/serokell/edna/blob/mkaito/edna14-oci-images/docker.nix#L38

File: servers/castor/edna.nix
Comment (mkaito) [reply servers/castor/edna.nix:87]: Oh that's fine, I just put it here as a placeholder. I'll change it to whatever we're actually using once we start using it. I don't mind separate options or conn string.

Upstream postgres container requires setting some environment variables to control db initialization: https://github.com/serokell/gemini-infra/blob/mkaito/edna14-castor/servers/castor/edna.nix#L115

File: servers/castor/edna.nix
Comment (mkaito) [reply servers/castor/edna.nix:78]: There's a thing or two I could do, but I wouldn't worry too much. I'm aware the config structure is unstable, and I expect that I'll have to change things soon. You can consider the entire construct a placeholder.

This is actually one reason I asked about config merging. It would be ideal for me to only specify options that differ from their defaults.

File: servers/castor/edna.nix
Comment (gromakovsky) [reply servers/castor/edna.nix:87]: Okay, then I think we'll merge it with `conn-string` option and it will have to be updated here. https://github.com/serokell/edna/pull/8 requires just 3 tiny fixes, and if @sashasashasasha151 shows up today, most likely we'll merge it today. Let's update this place after we merge we merge https://github.com/serokell/edna/pull/8 to be sure we won't change it.

File: servers/castor/edna.nix
Comment (gromakovsky) [reply servers/castor/edna.nix:78]: Ok, probably after https://github.com/serokell/edna/pull/8 we will not make config changes for a while (but it's hard to predict), so let's wait for it to be merged, update the config here and stop worrying for now.
AFAIU, if we update the config structure so that the old config no longer can be parsed and merge it, **master** CI build will try to redeploy staging and will fail until we update this place in this repo, right? That's not critical at this stage, just asking.

> This is actually one reason I asked about config merging. It would be ideal for me to only specify options that differ from their defaults.

Yes, I think we will implement config merging and reading from env variables and command line options at some point, it's just a matter of priorities. Currently we have a lot of other work and tight deadlines, so we just provide _some way_ to configure `edna-server` which is not perfect. I'll create an issue about that, but it will stay in backlog for a while.

File: flake.nix
Comment (gromakovsky) [inline flake.nix:5]: Btw, it's not my business, but I see it's approved, so can it be merged and can master be used here?

File: servers/castor/edna.nix
Comment (gromakovsky) [reply servers/castor/edna.nix:78]: `edna-server` changes have been merged, see https://github.com/serokell/edna/pull/11#issuecomment-799206708

Comment (gromakovsky) [reply servers/castor/edna.nix:78]: > I'll create an issue about that, but it will stay in backlog for a while.

https://issues.serokell.io/issue/EDNA-44

File: servers/castor/edna.nix
Comment (gromakovsky) [inline servers/castor/edna.nix:87]: It should be inside `db`.

**Review by gromakovsky — APPROVED** [review]
I approve as long as you make sure that the config defined here matches the format we expect. I think we'll rename `max-connection` and `initiation` as you suggested.

File: servers/castor/edna.nix
Comment (mkaito) [reply servers/castor/edna.nix]: Oh. Ok, fixed.

**Pull request merged**
